### PR TITLE
Added MGR recipe for thermal compositional solver

### PR DESCRIPTION
**Pull request by francoishamon**
This PR adds an MGR recipe for the thermal compositional case. 

So far, I used the CPTR approach:
- Eliminate the last density, aligned with volume balance
- Eliminate the remaining densities
- On the coarse grid, solve for pressure and temperature using AMG with `HYPRE_BoomerAMGSetNumFunctions( mgrData.coarseSolver.ptr, 2 )`.

I have something that works on large problems and that seems to scale, so we can merge it now and improve it later.

### CODE REVIEW

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp [hunk 1]
```diff
@@ -50,14 +50,21 @@ CompositionalMultiphaseFVM::CompositionalMultiphaseFVM( const string & name,
...
 
+  if( m_isThermal )
+  {
+    m_linearSolverParameters.get().mgr.strategy = LinearSolverParameters::MGR::StrategyType::thermalCompositionalMultiphaseFVM;
+  }
+  else
+  {
+    m_linearSolverParameters.get().mgr.strategy = LinearSolverParameters::MGR::StrategyType::compositionalMultiphaseFVM;
+  }
+
```
Comment (TotoGaz) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:67]:
```suggestion
  using ...;
  m_linearSolverParameters.get().mgr.strategy = m_isThermal ? StrategyType::thermalCompositionalMultiphaseFVM : StrategyType::compositionalMultiphaseFVM;

Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: Done

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalCompositionalMultiphaseFVM.hpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    m_levelCoarseGridMethod[0] = MGRCoarseGridMethod::galerkin; // Standard Galerkin
+
+    m_levelFRelaxMethod[1]     = MGRFRelaxationMethod::singleLevel; //default, i.e. Jacobi
+    m_levelInterpType[1]       = MGRInterpolationType::injection; // Injection
+    m_levelRestrictType[1]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[1] = MGRCoarseGridMethod::cprLikeBlockDiag; // Non-Galerkin Quasi-IMPES CPR
+
+    // Global smoother at each level, only do block-GS for the condensed system
+    m_levelSmoothType[1] = 1;
+    m_levelSmoothIters[1] = 1;
```
Comment (castelletto1) [inline src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalCompositionalMultiphaseFVM.hpp:81]: I probably would rephrase the comment. I think the sense is that global smoothing is performed only on the first level, after reservoir density associated with the volume constraint have been eliminated. Although it is the default option (no global smoothing), we may add for clarity:
```c++
m_levelSmoothIters[0] = 0;
```

Not in this PR, but we should add enumerations to account for new capabilities introduced in #1712 after updating _hypre_.

**Review by castelletto1 — APPROVED** [review]
Great, looks good to me!

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalCompositionalMultiphaseFVM.hpp (hunk 2 above)
Comment (francoishamon) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalCompositionalMultiphaseFVM.hpp]: Yes, I added the `m_levelSmoothIters[0] = 0;` to make it clear that there is no global smoothing in this case. And yes, I agree that we should update the enumerations after the hypre update. I started adding some of them in `HypreUtils.hpp` in the last commit of #1712, but this should probably be checked again / made more uniform.

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/CompositionalMultiphaseFVM.hpp [hunk 3]
```diff
@@ -101,7 +101,8 @@ class CompositionalMultiphaseFVM : public MGRStrategyBase< 2 >
     GEOSX_LAI_CHECK_ERROR( HYPRE_MGRSetLevelSmoothIters( precond.ptr, m_levelSmoothIters ) );
     GEOSX_LAI_CHECK_ERROR( HYPRE_MGRSetTruncateCoarseGridThreshold( precond.ptr, 1e-20 )); // truncate intermediate/coarse grids
 
-    // Note for me: uncomment HYPRE_MGRSetLevelFRelaxMethod and comment HYPRE_MGRSetRelaxType breaks the recipe
+    // Note: uncommenting HYPRE_MGRSetLevelFRelaxMethod and commenting HYPRE_MGRSetRelaxType breaks the recipe. This requires further
+    // investigation
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/CompositionalMultiphaseFVM.hpp:105]: I was confused again by the difference between the `FRelaxMethod` and the `FRelaxType`. I realized this afternoon that when I specify both the `Method` and `Type`, then everything works fine (which is not the case when I do not specify the `Type`). So, in the latest commit, I removed this comment.

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalCompositionalMultiphaseFVM.hpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    GEOSX_LAI_CHECK_ERROR( HYPRE_MGRSetNumRelaxSweeps( precond.ptr, 1 ));
+
+#ifdef GEOSX_USE_HYPRE_CUDA
+    GEOSX_LAI_CHECK_ERROR( HYPRE_MGRSetRelaxType( precond.ptr, getAMGRelaxationType( LinearSolverParameters::AMG::SmootherType::l1jacobi ) ) );
+#endif
+
+    GEOSX_LAI_CHECK_ERROR( HYPRE_BoomerAMGCreate( &mgrData.coarseSolver.ptr ) );
+    GEOSX_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetPrintLevel( mgrData.coarseSolver.ptr, 0 ) );
+    GEOSX_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetMaxIter( mgrData.coarseSolver.ptr, 1 ) );
+    GEOSX_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetAggNumLevels( mgrData.coarseSolver.ptr, 1 ) );
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalCompositionalMultiphaseFVM.hpp:120]: Yes, after your comments I tested on a thermal unstructured case with about 260k cells by commenting out `GEOSX_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetAggNumLevels( mgrData.coarseSolver.ptr, 1 ) );`.

With aggressive coarsening:
Run time: 1061.098s

Without aggressive coarsening:
Reduction in the number of linear iterations by about 30% 
Run time: 971.072s 

so yes, I will remove aggressive coarsening
